operator/cache_memcached.go

```diff
@@ -220,3 +279,60 @@ func (m *MemcachedStore) getKey(obj any) (prefixedKey string, externalKey string
 	}
 	return fmt.Sprintf("%s/%s", m.kind.Plural(), externalKey), externalKey, nil
 }
+
+func (m *MemcachedStore) setKeysFromCache() error {
+	item, err := m.client.Get(fmt.Sprintf(keysCacheKey, m.kind.Plural()))
+	if err != nil {
+		if errors.Is(err, memcache.ErrCacheMiss) {
+			return nil
+		}
+		return err
+	}
+	keys := make([]string, 0)
+	err = json.Unmarshal(item.Value, &keys)
+	if err != nil {
+		return err
+	}
+	for _, key := range keys {
+		m.keys.Store(key, struct{}{})
+	}
+	return nil
+}
+
+func (m *MemcachedStore) syncKeys() error {
```

Comment:
IIUC this is doing a two-way sync with merge, which can get hairy in terms of consistency.

Personally I'd prefer something like this:

1. Always write to in-mem list of keys first.
2. Sync from in-mem to memcached periodically.
3. `ListKeys` returns in-mem data.
4. When starting up, load initial vals from memcached.

That way we only have one-way sync (local to memcached) most of the time and only reload data from memcached upon restart – this make it simpler (albeit somewhat less efficient).

One other thing that came to my mind is that we should probably prefix the keys key with the process ID of some sorts, so that we don't overwrite the value of other processes.

Reply:
So there are a couple of hairy things around this (which is part of why I didn't initially use the cache for keys). If memcached is restarted/cleared, the keys returned by `ListKeys` should also be empty, otherwise the informer gets upset. So we do the two-way sync only in the event that the key for the keys has been deleted. This is a scenario that _can't_ happen in the in-memory cache that the informer is tailored for. I can check _how_ upset the informer gets (IIRC it doesn't panic, so that might be "good enough") and adjust this to one-way with the caveat that restarting memcached may result in weird behavior and/or upset log messages.

For key prefixing, I was thinking of that with respect to horizontal scaling as well, given that the keys should be only the keys for the particular shard, so it doesn't try to resync (via `ListKeys`) on all of them. Process ID might not be the best approach because it can't sync _from_ the cache properly on startup if the process ID changes, but perhaps we add a "name" or "operator/shard ID" field which is required when building the cache which we can use to prefix.

Reply:
OK, so in case of memcached restarts, how about this – we do as I suggested in step 1-4 but modify [2] to also check for empty (i.e. restarted) keys before overwriting? In other words, keep the sync code minus the merging part and.

RE: process ID – `operator + shardID` is what I had in mind when saying process ID – sorry for the confusion 😅 

WDYT?

Reply:
ok, so testing results: if the key is returned by `ListKeys()`, but it doesn't exist in the cache, we get a `klog` error like:
```
I0726 16:22:13.377619       1 delta_fifo.go:730] Key default/foo does not exist in known objects store, unable to queue object for sync
```
On an update when the cache has been cleared, it comes through as an Add, as it's not in the cache (though, with OpinionatedWatcher, it'll be a Sync), much like the results of the initial list.

Interesting outcome of this testing: if the cache returns an error from `Get()`, the informer swallows it and the event is never propagated (no error is printed either). I'll have to dig in deeper as to what is happening there.